pkg/csi/recover/recover.go

```diff
@@ -0,0 +1,201 @@
+		return
+	}
+
+	for _, point := range brokenMounts {
+		glog.V(4).Infof("Get broken mount point: %v", point)
+		r.umountDuplicate(point)
+		if err := r.recoverBrokenMount(point); err != nil {
+			r.eventRecord(point, corev1.EventTypeWarning, common.FuseRecoverFailed)
+			continue
+		}
+		r.eventRecord(point, corev1.EventTypeNormal, common.FuseRecoverSucceed)
+	}
+}
+
+func (r *FuseRecover) recoverBrokenMount(point mountinfo.MountPoint) (err error) {
+	glog.V(3).Infof("Start recovery: [%s], source path: [%s]", point.MountPath, point.SourcePath)
+	// recovery for each bind mount path
+	mountOption := []string{"bind"}
+	if point.ReadOnly {
+		mountOption = append(mountOption, "ro")
+	}
+
+	glog.V(3).Infof("Start exec cmd: mount %s %s -o %v \n", point.SourcePath, point.MountPath, mountOption)
+	if err := r.Mount(point.SourcePath, point.MountPath, "none", mountOption); err != nil {
+		glog.Errorf("exec cmd: mount -o bind %s %s err :%v", point.SourcePath, point.MountPath, err)
+	}
+	return
+}
+
+// check mountpoint count
+// umount duplicate mountpoint util 1 avoiding very large mountinfo file.
+// don't umount all item, 'mountPropagation' will lose efficacy.
+func (r *FuseRecover) umountDuplicate(point mountinfo.MountPoint) {
+	for i := point.Count; i > 1; i-- {
+		glog.V(3).Infof("count: %d, start exec cmd: umount %s", i, point.MountPath)
+		if err := r.Unmount(point.MountPath); err != nil {
+			glog.Errorf("exec cmd: umount %s err: %v", point.MountPath, err)
+		}
+	}
+}
+
+func (r *FuseRecover) compareOrRecordContainerStat(pod corev1.Pod) (restarted bool) {
+	if pod.Status.ContainerStatuses == nil || len(pod.OwnerReferences) == 0 {
+		return
+	}
+	var dsName string
+	for _, obj := range pod.OwnerReferences {
+		if obj.Kind == "DaemonSet" {
+			dsName = obj.Name
+		}
+	}
+	if dsName == "" {
+		return
+	}
+	for _, cn := range pod.Status.ContainerStatuses {
+		if cn.State.Running == nil {
+			continue
+		}
+		key := fmt.Sprintf("%s-%s-%s", cn.Name, dsName, pod.Namespace)
+		cs, ok := r.containers[key]
+		if !ok {
+			cs = &containerStat{
+				name:          cn.Name,
+				podName:       pod.Name,
+				namespace:     pod.Namespace,
+				daemonSetName: dsName,
+				startAt:       cn.State.Running.StartedAt,
+			}
+			r.containers[key] = cs
+			continue
+		}
+
+		if cs.startAt.Before(&cn.State.Running.StartedAt) {
+			glog.Infof("Container %s of pod %s in namespace %s start time is %s, but record %s",
+				cn.Name, pod.Name, pod.Namespace, cn.State.Running.StartedAt.String(), cs.startAt.String())
+			r.containers[key].startAt = cn.State.Running.StartedAt
+			restarted = true
+			return
+		}
+	}
+	return
+}
+
+func (r *FuseRecover) eventRecord(point mountinfo.MountPoint, eventType, eventReason string) {
+	namespacedName := point.NamespacedDatasetName
+	strs := strings.Split(namespacedName, "-")
+	if len(strs) < 2 {
+		glog.V(3).Infof("can't parse dataset from namespacedName: %s", namespacedName)
+		return
+	}
+	namespace, datasetName := strs[0], strs[1]
```

Comment:
1. Dataset name may contains "-", e.g. `imagenet-dataset-beijing`, so simply using "-" as delimiter is not guaranteed to get associated namespace and name.

Here's a [workaround](https://github.com/fluid-cloudnative/fluid/blob/master/pkg/csi/fuse/nodeserver.go#L339) that resolves similar problem in Fluid CSI Plugin. Hope it does some help to you.